File: pkg-r/tests/testthat/test-citation-scan.R
test_that("parse_commons_citation matches the shared cases", {
  cases <- shared_fixture("citations")$parse_commons_citation$cases
  expect_gt(length(cases), 0)

  for (case in cases) {
    out <- parse_commons_citation(case$body)
    if (is.null(case$expected)) {
      expect_null(out, info = case$name)
    } else {
      expect_identical(
        out,
        list(explanation = case$expected$explanation, quote = case$expected$quote),
        info = case$name
      )
    }
  }
})

test_that("citation decisions match the shared record shape", {
  # The shape is what the trajectory reviewer reads back out of the span, so
  # drive the real producers rather than asserting on a hand-built list.
  cases <- shared_fixture("citations")$citation_decision$cases
  expect_setequal(
    vapply(cases, function(case) case$decision$status, character(1)),
    c("accepted", "rejected", "malformed")
  )

  malformed <- "<commons-citation>\n\nno blockquote\n\n</commons-citation>"

  for (case in cases) {
    want <- case$decision
    got <- if (identical(want$status, "malformed")) {
      project_citation_text(malformed, list())$decisions[[1]]
    } else {
      corpus <- if (identical(want$status, "accepted")) {
        list(list(label = want$label, kind = want$kind, text = want$quote))
      } else {
        list(list(label = "documentation", kind = "prose", text = "No support."))
      }
      render_citation_aside(want$quote, "why", corpus)$decision
    }
    expect_identical(
      jsonlite::fromJSON(
        jsonlite::toJSON(got, auto_unbox = TRUE),
        simplifyVector = FALSE
      ),
      want,
      info = case$name
    )
  }
})

scan_chunks <- function(chunks, corpus) {
  scanner <- citation_scanner(corpus)
  fed <- vapply(chunks, scanner$feed, character(1), USE.NAMES = FALSE)
  list(
    text = paste0(paste(fed, collapse = ""), scanner$finish()),
    decisions = scanner$decisions()
  )
}

# {{pad}} keeps the cap cases small on disk: the fixture states the body length
# it wants and the runner builds the filler.
citation_scan_padding <- function(case, text) {
  if (is.null(case$pad)) {
    return("")
  }
  after_open <- sub(".*?<commons-citation>", "", text)
  body <- sub("</commons-citation>.*", "", after_open)
  strrep(
    case$pad$char,
    case$pad$citation_body_length - nchar(body) + nchar("{{pad}}")
  )
}

# The two packages resolve different icon URLs, so the fixture leaves each
# aside to the package reading it and pins every other character exactly.
citation_scan_expected <- function(case, corpus, padding) {
  expected <- gsub("{{pad}}", padding, case$expected_text, fixed = TRUE)
  for (i in seq_along(case$asides)) {
    aside <- case$asides[[i]]
    html <- render_citation_aside(
      aside$quote,
      gsub("{{pad}}", padding, aside$explanation, fixed = TRUE),
      corpus
    )$html
    expected <- gsub(
      paste0("{{aside:", i - 1L, "}}"),
      html,
      expected,
      fixed = TRUE
    )
  }
  expected
}

# Chunk invariance is a property of every case, not a case of its own, so each
# one is fed whole, at the fixture's own boundaries, and at every boundary a
# short text has.
citation_scan_chunkings <- function(text, chunks, exhaustive_max) {
  chunkings <- list("the whole text" = text)
  if (length(chunks) > 1) {
    chunkings[["the fixture's feed boundaries"]] <- chunks
  }
  if (nchar(text) <= exhaustive_max) {
    chars <- strsplit(text, "", fixed = TRUE)[[1]]
    chunkings[["one character at a time"]] <- chars
    for (at in seq_len(nchar(text) - 1L)) {
      chunkings[[paste("a split at", at)]] <- c(
        substr(text, 1, at),
        substr(text, at + 1L, nchar(text))
      )
    }
  }
  chunkings
}

as_json_list <- function(x) {
  jsonlite::fromJSON(
    jsonlite::toJSON(x, auto_unbox = TRUE),
    simplifyVector = FALSE
  )
}

test_that("the streaming scanner matches the shared cases", {
  spec <- shared_fixture("citations")$citation_scan
  expect_gte(length(spec$cases), 20)
  expect_identical(ELEMENT_BODY_CAP, as.integer(spec$element_body_cap))

  for (case in spec$cases) {
    corpus <- spec$corpora[[case$corpus]]
    text <- gsub("{{split}}", "", case$text, fixed = TRUE)
    padding <- citation_scan_padding(case, text)
    text <- gsub("{{pad}}", padding, text, fixed = TRUE)
    chunks <- strsplit(
      gsub("{{pad}}", padding, case$text, fixed = TRUE),
      "{{split}}",
      fixed = TRUE
    )[[1]]
    expected <- citation_scan_expected(case, corpus, padding)

    chunkings <- citation_scan_chunkings(
      text,
      chunks,
      spec$exhaustive_split_max_chars
    )
    for (how in names(chunkings)) {
      got <- scan_chunks(chunkings[[how]], corpus)
      info <- paste0(case$name, ", fed ", how)
      expect_identical(got$text, expected, info = info)
      expect_identical(as_json_list(got$decisions), case$decisions, info = info)
    }
  }
})

test_that("each feed emits the shared held-back cases", {
  cases <- shared_fixture("citations")$citation_holdback$cases
  expect_gt(length(cases), 0)

  corpus <- list(list(
    label = "documentation",
    kind = "prose",
    text = "Canopy cover is always acre-weighted for reporting."
  ))

  for (case in cases) {
    scanner <- citation_scanner(corpus)
    emitted <- vapply(
      case$chunks,
      function(chunk) scanner$feed(chunk),
      character(1),
      USE.NAMES = FALSE
    )
    expect_identical(as.list(emitted), case$emitted, info = case$name)
    expect_identical(scanner$finish(), case$flushed, info = case$name)
  }
})

test_that("an oversized unclosed citation keeps only a bounded close-tag suffix", {
  # Memory, not projection: the shared cases pin that nothing is emitted, and
  # this pins that nothing is retained while waiting for a close tag.
  corpus <- list(list(
    label = "documentation",
    kind = "prose",
    text = "Canopy cover is always acre-weighted for reporting."
  ))
  scanner <- citation_scanner(corpus)

  expect_identical(
    scanner$feed(paste0(
      "<commons-citation>",
      strrep("x", ELEMENT_BODY_CAP + 1L)
    )),
    ""
  )
  expect_lte(
    nchar(get("buf", envir = environment(scanner$feed))),
    nchar(CITATION_CLOSE) - 1L
  )
  expect_identical(scanner$feed(strrep("y", ELEMENT_BODY_CAP * 2L)), "")
  expect_lte(
    nchar(get("buf", envir = environment(scanner$feed))),
    nchar(CITATION_CLOSE) - 1L
  )
  expect_identical(scanner$finish(), "")
})

File: pkg-py/tests/test_citation_scan.py
"""The streaming scanner that projects model text for display.

The projection is a cross-language contract, so the cases live in
``tests/shared/citations.json`` under ``citation_scan`` and the R suite runs
the same ones. Do not restate a case here; add it to the fixture.
"""

from collections.abc import Iterator, Sequence
from typing import Any

import pytest

from commons._citation_scan import CITATION_CLOSE, ELEMENT_BODY_CAP, CitationScanner
from commons._citations import CorpusEntry, citation_aside_html, match_citation

from ._shared import load_shared_fixture

SPEC: dict[str, Any] = load_shared_fixture("citations")["citation_scan"]
CASES: list[dict[str, Any]] = SPEC["cases"]
HOLDBACK_CASES: list[dict[str, Any]] = load_shared_fixture("citations")[
    "citation_holdback"
]["cases"]
CORPORA: dict[str, list[dict[str, Any]]] = SPEC["corpora"]
EXHAUSTIVE_MAX: int = SPEC["exhaustive_split_max_chars"]

PAD_TOKEN = "{{pad}}"
SPLIT_TOKEN = "{{split}}"
CITATION_OPEN = "<commons-citation>"

# One sample passage for the tests below that need a quote but do not pin the
# projection. Anything that pins the projection belongs in the fixture.
SAMPLE_QUOTE = "Canopy cover is always acre-weighted for reporting."
SAMPLE_CORPUS = [CorpusEntry(label="documentation", kind="prose", text=SAMPLE_QUOTE)]


def scan(chunks: Sequence[str], corpus: Sequence[CorpusEntry]) -> tuple[str, list[Any]]:
    scanner = CitationScanner(corpus)
    text = "".join(scanner.feed(chunk) for chunk in chunks) + scanner.finish()
    return text, [decision.as_record() for decision in scanner.candidates]


def _padding(case: dict[str, Any], text: str) -> str:
    """Expand ``{{pad}}`` so the first citation body reaches its stated length."""
    pad = case.get("pad")
    if pad is None:
        return ""
    body = text.split(CITATION_OPEN, 1)[1].split(CITATION_CLOSE, 1)[0]
    return pad["char"] * (pad["citation_body_length"] - len(body) + len(PAD_TOKEN))


def _render(
    case: dict[str, Any],
    expected: str,
    corpus: Sequence[CorpusEntry],
    padding: str,
) -> str:
    """Substitute each ``{{aside:N}}`` with the aside this package renders."""
    for index, aside in enumerate(case.get("asides", [])):
        quote = aside["quote"]
        entry = match_citation(quote, corpus)
        assert entry is not None, f"aside {index} of {case['name']} does not verify"
        explanation = aside["explanation"].replace(PAD_TOKEN, padding)
        html = citation_aside_html(quote, explanation, entry.label, entry.kind)
        expected = expected.replace(f"{{{{aside:{index}}}}}", html)
    return expected


def _chunkings(text: str, explicit: list[str]) -> Iterator[tuple[str, list[str]]]:
    yield "the whole text", [text]
    if len(explicit) > 1:
        yield "the fixture's feed boundaries", explicit
    if len(text) <= EXHAUSTIVE_MAX:
        yield "one character at a time", list(text)
        for at in range(1, len(text)):
            yield f"a split at {at}", [text[:at], text[at:]]


def test_shared_fixture_is_populated() -> None:
    # Parametrizing over an empty list collects nothing and the suite still
    # passes, so pin what the fixture has to cover.
    assert len(CASES) >= 20
    statuses = {decision["status"] for case in CASES for decision in case["decisions"]}
    assert statuses == {"accepted", "rejected", "malformed"}


def test_the_cap_matches_the_shared_fixture() -> None:
    assert ELEMENT_BODY_CAP == SPEC["element_body_cap"]


@pytest.mark.parametrize("case", CASES, ids=lambda case: case["name"])
def test_projection_matches_the_shared_cases(case: dict[str, Any]) -> None:
    corpus = [CorpusEntry(**entry) for entry in CORPORA[case["corpus"]]]
    text = case["text"].replace(SPLIT_TOKEN, "")
    padding = _padding(case, text)
    text = text.replace(PAD_TOKEN, padding)
    chunks = case["text"].replace(PAD_TOKEN, padding).split(SPLIT_TOKEN)
    expected = _render(
        case, case["expected_text"].replace(PAD_TOKEN, padding), corpus, padding
    )

    for description, fed in _chunkings(text, chunks):
        got_text, got_decisions = scan(fed, corpus)
        assert got_text == expected, f"fed {description}"
        assert got_decisions == case["decisions"], f"fed {description}"


@pytest.mark.parametrize("case", HOLDBACK_CASES, ids=lambda case: case["name"])
def test_each_feed_emits_the_shared_cases(case: dict[str, Any]) -> None:
    scanner = CitationScanner(SAMPLE_CORPUS)

    emitted = [scanner.feed(chunk) for chunk in case["chunks"]]

    assert emitted == case["emitted"]
    assert scanner.finish() == case["flushed"]


def test_any_verified_reports_whether_a_quote_was_accepted() -> None:
    scanner = CitationScanner(SAMPLE_CORPUS)
    assert not scanner.any_verified

    scanner.feed("<commons-citation>\n\nwhy\n\n> nothing supports this\n\n")
    scanner.feed("</commons-citation>")
    assert not scanner.any_verified

    scanner.feed(f"\n<commons-citation>\n\nwhy\n\n> {SAMPLE_QUOTE}\n\n")
    scanner.feed("</commons-citation>")
    assert scanner.any_verified


def test_an_oversized_unclosed_citation_does_not_grow_the_buffer() -> None:
    # Memory, not projection: the fixture pins that nothing is emitted, and
    # this pins that nothing is retained while waiting for a close tag.
    scanner = CitationScanner(SAMPLE_CORPUS)

    assert scanner.feed("<commons-citation>" + "x" * (ELEMENT_BODY_CAP + 1)) == ""
    assert len(scanner._buffer) < len(CITATION_CLOSE)
    assert scanner.feed("y" * (ELEMENT_BODY_CAP * 2)) == ""
    assert len(scanner._buffer) < len(CITATION_CLOSE)
    assert scanner.finish() == ""


def test_the_aside_escapes_its_label_and_reflows_the_quote() -> None:
    html = citation_aside_html(
        "first line\nsecond line", "why", 'a "quoted" & label', "prose"
    )

    # The attribute and the title text sit in different contexts, so they are
    # escaped differently: quotes in the attribute, angle brackets in the text.
    assert html.startswith('<shiny-aside label="a &quot;quoted&quot; &amp; label">')
    assert '>a "quoted" &amp; label</span>' in html
    assert html.endswith("why\n\n> first line\n> second line</shiny-aside>")


def test_the_aside_omits_the_explanation_when_the_model_gave_none() -> None:
    html = citation_aside_html(SAMPLE_QUOTE, "", "documentation", "prose")

    assert html.endswith(f"</span>\n\n> {SAMPLE_QUOTE}</shiny-aside>")
